client/retry/backoff.go

```diff
@@ -0,0 +1,87 @@
+// Copyright 2023 TiKV Project Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package retry
+
+import (
+	"context"
+	"time"
+
+	"github.com/pingcap/failpoint"
+)
+
+// BackOffer is a backoff policy for retrying operations.
+type BackOffer struct {
+	maxBackoff     time.Duration
+	nextBackoff    time.Duration
+	initialBackoff time.Duration
+}
+
+// WithBackoff is a helper function to add backoff.
+func WithBackoff(
+	ctx context.Context,
+	fn func() error,
+	bo *BackOffer,
+) error {
+	if err := fn(); err != nil {
+		select {
+		case <-ctx.Done():
+		case <-time.After(bo.NextBackoff()):
+			failpoint.Inject("backOffExecute", func() {
+				testBackOffExecuteFlag = true
+			})
+		}
+		return err
+	}
+	// Reset backoff when fn() succeed.
+	bo.ResetBackoff()
+	return nil
+}
+
+// InitialBackOffer make the initial state for retrying.
+func InitialBackOffer(initialBackoff, maxBackoff time.Duration) BackOffer {
+	return BackOffer{
+		maxBackoff:     maxBackoff,
+		initialBackoff: initialBackoff,
+		nextBackoff:    initialBackoff,
+	}
+}
+
+// NextBackoff implements the `Backoffer`, for now use the `ExponentialBackoff`.
+func (rs *BackOffer) NextBackoff() time.Duration {
+	return rs.exponentialBackoff()
+}
+
+// ExponentialBackoff Get the exponential backoff duration.
+func (rs *BackOffer) exponentialBackoff() time.Duration {
```

Comment:
Will we use exponential backoff by default? When network partition a long time, it may need more time to recover.

Reply:
limited by max backoff time...

Reply:
how long is it?